packages/middleware-retry/src/defaultStrategy.ts

```diff
@@ -33,46 +34,78 @@ export interface DelayDecider {
   (delayBase: number, attempts: number): number;
 }
 
+/**
+ * Interface that specifies the retry quota behavior.
+ */
+export interface RetryQuota {
+  /**
+   * returns true if retry tokens are available from the retry quota bucket.
+   */
+  hasRetryTokens: (error: SdkError) => boolean;
+
+  /**
+   * returns token amount from the retry quota bucket.
+   * throws error is retry tokens are not available.
+   */
+  retrieveRetryTokens: (error: SdkError) => number;
+
+  /**
+   * releases tokens back to the retry quota.
+   */
+  releaseRetryTokens: (releaseCapacityAmount?: number) => void;
+}
+
 /**
  * Strategy options to be passed to StandardRetryStrategy
  */
 export interface StandardRetryStrategyOptions {
   retryDecider?: RetryDecider;
   delayDecider?: DelayDecider;
+  retryQuota?: RetryQuota;
```

Comment:
This `retryQuota` config interface doesn't look straight forward enough as a constructor parameter. If we are going to make it configurable, we should allow changing the parameters instead of the behavior(`hasRetryTokens`, `retrieveRetryTokens` and `releaseRetryTokens` functions). For example, we can make `initialRetryTokens` configurable. The `RetryQuota` interface can be initiated internally.

Reply:
In the existing solution, if customer wants to override `RetryQuota` interface they need to implement the methods `hasRetryTokens`, `retrieveRetryTokens` and `releaseRetryTokens` in whatever way they please.

We can pass parameters in the constructor to make it more strict. If yes, which ones should we pass?

Options am in favor of:
* INITIAL_RETRY_TOKENS

Options am not in favor of (as they're very specific to default RetryQuota implementation):
* TIMEOUT_RETRY_COST
* RETRY_COST
* NO_RETRY_INCREMENT

Reply:
Agreement after offline discussion:
* retain existing functions for RetryQuota interface: `hasRetryTokens`, `retrieveRetryTokens` and `releaseRetryTokens`
* make INITIAL_RETRY_TOKENS configurable in defaultRetryQuota

Reply:
Updated in [commit](https://github.com/aws/aws-sdk-js-v3/pull/1258/commits/fb44543e63b43873977c8b1a7cbef211e9c724db)